crate/pkcs11/provider/src/kms_object.rs

```diff
@@ -0,0 +1,170 @@
+                let other_tags = attributes
+                    .get_tags()
+                    .into_iter()
+                    .filter(|t| !(t.is_empty() || tags.contains(t) || t.starts_with('_')))
+                    .collect::<Vec<String>>();
+                results.push(KmsObject {
+                    object: object.clone(),
+                    attributes: attributes.clone(),
+                    other_tags,
+                });
+            }
+            Err(e) => {
+                return Err(Pkcs11Error::ServerError(e.to_string()));
+            }
+        }
+    }
+    Ok(results)
+}
+
+async fn locate_objects(
+    kms_client: &KmsClient,
+    tags: &[String],
+) -> Result<Vec<String>, Pkcs11Error> {
+    let mut attributes = Attributes::default();
+    attributes.set_tags(tags)?;
+
+    let locate = Locate {
+        maximum_items: None,
+        offset_items: None,
+        storage_status_mask: None,
+        object_group_member: None,
+        attributes,
+    };
+    let response = kms_client.locate(locate).await?;
+    let uniques_identifiers = response
+        .unique_identifiers
+        .unwrap_or(vec![])
+        .iter()
+        .map(|id| id.to_string().unwrap_or_default())
+        .filter(|id| !id.is_empty())
+        .collect();
+    debug!(
+        "Located objects: tags: {:?} => {:?}",
+        tags, uniques_identifiers
+    );
+    Ok(uniques_identifiers)
+}
+
+pub fn kms_decrypt(
+    kms_client: &KmsClient,
+    key_id: &str,
+    encryption_algorithm: EncryptionAlgorithm,
+    data: &[u8],
+) -> Result<Zeroizing<Vec<u8>>, Pkcs11Error> {
+    tokio::runtime::Runtime::new()?.block_on(kms_decrypt_async(
+        kms_client,
+        key_id,
+        encryption_algorithm,
+        data,
+    ))
+}
+
+pub(crate) async fn kms_decrypt_async(
+    kms_client: &KmsClient,
+    key_id: &str,
+    encryption_algorithm: EncryptionAlgorithm,
+    data: &[u8],
+) -> Result<Zeroizing<Vec<u8>>, Pkcs11Error> {
+    let cryptographic_parameters = match encryption_algorithm {
+        EncryptionAlgorithm::RsaPkcs1v15 => CryptographicParameters {
+            cryptographic_algorithm: Some(CryptographicAlgorithm::RSA),
+            padding_method: Some(PaddingMethod::PKCS1v15),
+            ..Default::default()
+        },
+        // _ => {
+        //     return Err(Pkcs11Error::ServerError(format!(
+        //         "Unsupported encryption algorithm: {:?}",
+        //         encryption_algorithm
+        //     )));
+        // }
+    };
+    let decryption_request = Decrypt {
+        unique_identifier: Some(UniqueIdentifier::TextString(key_id.to_owned())),
+        cryptographic_parameters: Some(cryptographic_parameters),
+        data: Some(data.to_vec()),
+        iv_counter_nonce: None,
+        correlation_value: None,
+        init_indicator: None,
+        final_indicator: None,
+        authenticated_encryption_additional_data: None,
+        authenticated_encryption_tag: None,
+    };
```

Comment:
OK for Default()
Taking ownership: I reworked the API because, yes, when you decrypt you usually do not use the ciphertext anymore... but at the end of the day, it soes not save an allocation because we have to take ownership of a slice from the C API